firefox-ios/Client/Frontend/Home/Homepage Rebuild/TopSites/TopSitesManager.swift

```diff
@@ -3,30 +3,72 @@
 // file, You can obtain one at http://mozilla.org/MPL/2.0/
 
 import Foundation
+import Common
 import Shared
 import Storage
 
 // TODO: FXIOS-10165 - Add full logic + tests for retrieving top sites
+/// Manager to fetch the top sites data, The data gets updated from notifications on specific user actions
 class TopSitesManager {
-    private let googleTopSiteManager: GoogleTopSiteManager
+    private var logger: Logger
+    private let prefs: Prefs
+    private let contileProvider: ContileProviderInterface
+    private let googleTopSiteManager: GoogleTopSiteManagerProvider
 
     init(
-        googleTopSiteManager: GoogleTopSiteManager
+        prefs: Prefs,
+        contileProvider: ContileProviderInterface = ContileProvider(),
+        googleTopSiteManager: GoogleTopSiteManagerProvider,
+        logger: Logger = DefaultLogger.shared
     ) {
+        self.prefs = prefs
+        self.contileProvider = contileProvider
         self.googleTopSiteManager = googleTopSiteManager
+        self.logger = logger
     }
 
     func getTopSites() async -> [TopSiteState] {
-        guard let googleTopSite = addGoogleTopSite() else {
+        var topSites: [TopSiteState] = []
+
+        let googleTopSite = addGoogleTopSite()
+        let sponsoredTopSites = await getSponsoredSites()
+
+        topSites = googleTopSite + sponsoredTopSites
+        return topSites
+    }
+
+    private func addGoogleTopSite() -> [TopSiteState] {
+        guard let googleSite = googleTopSiteManager.suggestedSiteData else {
             return []
         }
-        return [googleTopSite]
+        return [TopSiteState(site: googleSite)]
     }
 
-    private func addGoogleTopSite() -> TopSiteState? {
-        guard let googleSite = googleTopSiteManager.suggestedSiteData else {
-            return nil
+    // MARK: Sponsored tiles (Contiles)
+    private var shouldLoadSponsoredTiles: Bool {
+        return prefs.boolForKey(PrefsKeys.UserFeatureFlagPrefs.SponsoredShortcuts) ?? true
+    }
+
+    private func getSponsoredSites() async -> [TopSiteState] {
+        guard shouldLoadSponsoredTiles else { return [] }
+
+        let contiles = await withCheckedContinuation { continuation in
```

Comment:
there are some risks with using `withCheckedContinuation` from what I've read, but this seems to be the way to wrap completions to utilize modern concurrency and the codepaths here are simple as there are only two paths that can occur when resuming.

cc: @OrlaM

Reply:
I'm always weary of this, but you've covered all exit cases, so, it's fine.

Reply:
same, hopefully the tests will catch any issues too. thanks!